README.md
[![Coverage Status](https://coveralls.io/repos/lukashinsch/spring-properties-decrypter/badge.svg?branch=master)](https://coveralls.io/r/lukashinsch/spring-properties-decrypter?branch=master)
[![Build Status](https://travis-ci.org/lukashinsch/spring-properties-decrypter.svg?branch=master)](https://travis-ci.org/lukashinsch/spring-properties-decrypter)
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/eu.hinsch/spring-properties-decrypter/badge.svg)](https://maven-badges.herokuapp.com/maven-central/eu.hinsch/spring-properties-decrypter/)

# spring-properties-decrypter
Transparently decrypt property values to be used in spring environment using [jasypt](http://www.jasypt.org/) library.

# Howto use

## Gradle dependency
```
runtime('eu.hinsch:spring-properties-decrypter:0.1.8')
```

## Maven dependency
```
<dependency>
  <groupId>eu.hinsch</groupId>
  <artifactId>spring-properties-decrypter</artifactId>
  <version>0.1.8</version>
</dependency>
```

## Decryption password
Set decryption key via environment or system property
```
propertyDecryption.password=SECRETPASSWORD
```

## Encrpted properties
Define any spring environment property anywhere it can be defined (application*.properties/yaml, system, environment, command line...) and insert encrypted value like this:
```
myProperty={encrypted}ENCRYPTEDVALUERAWDATA
```

## Alternative prefix
To use a prefix other than '{encrypted}' define via
```
propertyDecryption.prefix=MY-PREFIX
```

## Encryption algorithm
By default uses PBEWithMD5AndDES (to allow running in default JRE). Note that this encryption is (at least with lots of dedicated resources) breakable. Stronger encryption is currently not working due to a bug in the underlying library.

To configure alternative algorithm:
```
propertyDecryption.algorithm=ALGORITHMNAME
```

## Howto encrypt passwords
Download the jasypt distribution from http://www.jasypt.org/download.html
Run from the bin folder
```
encrypt [verbose=true] algorithm=PBEWithMD5AndDES saltGeneratorClassName=org.jasypt.salt.ZeroSaltGenerator password=... input=...
```
For more details http://www.jasypt.org/cli.html


CHANGELOG.md
# Changelog

## 0.1.9
- Provide information on what property cannot be decrypted (fixes [#8](https://github.com/lukashinsch/spring-properties-decrypter/issues/8))

## 0.1.8
- Bugfix: application listener was not working for most types of property sources (fixes [#4](https://github.com/lukashinsch/spring-properties-decrypter/issues/4))

## 0.1.7
- Switched from PropertySourcesPlaceholderConfigurer to ApplicationListener to avoid possible conflicts

## 0.1.6
- propertyDecryption.prefix no longer removes itself (fixes [#2](https://github.com/lukashinsch/spring-properties-decrypter/issues/2))

## 0.1.5
- Change default prefix to {encrypted}

## 0.1.4
- Make algorithm configurable
- Don't use salt generator to decrypt passwords

## 0.1.3
- Relaxed spring dependency requirements a bit

## 0.1.2
- Clean up naming (decryptor -> decrypter)

## 0.1.1
- Add spring boot auto configuration

## 0.1.0
- Initial release

src/test/java/eu/hinsch/spring/propertiesdecrypter/DecryptingPropertiesApplicationListenerTest.java
package eu.hinsch.spring.propertiesdecrypter;

import org.jasypt.exceptions.EncryptionOperationNotPossibleException;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.context.event.ApplicationEnvironmentPreparedEvent;
import org.springframework.core.env.PropertiesPropertySource;
import org.springframework.core.env.StandardEnvironment;

import java.util.Properties;

import static org.hamcrest.Matchers.isA;

/**
 * Created by lh on 27/06/15.
 */
public class DecryptingPropertiesApplicationListenerTest {

    @Rule
    public ExpectedException exception = ExpectedException.none();

    private final DecryptingPropertiesApplicationListener listener = new DecryptingPropertiesApplicationListener();

    @Test
    public void shouldWrapExceptionOnEncryptionError() {
        // given
        StandardEnvironment environment = new StandardEnvironment();
        Properties props = new Properties();
        props.put("propertyDecryption.password", "NOT-A-PASSWORD");
        props.put("someProperty", "{encrypted}NOT-ENCRYPTED");
        environment.getPropertySources().addFirst(new PropertiesPropertySource("test-env", props));

        exception.expect(RuntimeException.class);
        exception.expectMessage("Unable to decrypt property value '{encrypted}NOT-ENCRYPTED'");
        exception.expectCause(isA(EncryptionOperationNotPossibleException.class));

        ApplicationEnvironmentPreparedEvent event = new ApplicationEnvironmentPreparedEvent(new SpringApplication(), new String[0], environment);

        // when
        listener.onApplicationEvent(event);

        // then -> exception
    }
}


src/test/java/eu/hinsch/spring/propertiesdecrypter/DecryptingPropertySourcesPlaceholderConfigurerTest.java
package eu.hinsch.spring.propertiesdecrypter;

import org.junit.Test;
import org.junit.experimental.runners.Enclosed;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.test.SpringApplicationConfiguration;
import org.springframework.core.env.Environment;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

/**
 * Created by lh on 04/04/15.
 */
@RunWith(Enclosed.class)
public class DecryptingPropertySourcesPlaceholderConfigurerTest {

    private static final String CODE = "7M2qVa5OHzn43YWGUE6R2Q==";
    private static final String ENCRYPTED_VALUE = "{encrypted}" + CODE;
    private static final String SECRET = "MY-SECRET";

    @SpringBootApplication
    static class TestAppConfig {
    }

    @TestPropertySource(properties = "secretProperty = " + ENCRYPTED_VALUE)
    public static class WithSecretProperty extends TestConfig {
        @Test
        public void shouldDecodeSystemProperty() throws Exception {
            assertThat(environment.getProperty("secretProperty"), is(SECRET));
        }
    }

    @TestPropertySource(properties = "conflictingProperty = " + ENCRYPTED_VALUE)
    public static class WithConflictingProperty extends TestConfig {
        @Test
        public void shouldDecryptCorrectPropertyOnConflict() throws Exception {
            assertThat(environment.getProperty("conflictingProperty"), is(SECRET));
        }
    }

    @TestPropertySource(properties = {
            "propertyDecryption.prefix = ---SOME-PREFIX---",
            "secretProperty = ---SOME-PREFIX---" + CODE
    })
    public static class WithPrefix extends TestConfig {
        @Test
        public void shouldUsePrefixOverride() {
            assertThat(environment.getProperty("secretProperty"), is(SECRET));
        }

        @Test
        public void shouldNotChangePrefixProperty() {
            assertThat(environment.getProperty("propertyDecryption.prefix"), is("---SOME-PREFIX---"));
        }
    }

    @DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
    @SpringApplicationConfiguration(classes = DecryptingPropertySourcesPlaceholderConfigurerTest.TestAppConfig.class)
    @RunWith(SpringJUnit4ClassRunner.class)
    @TestPropertySource(properties = "propertyDecryption.password = MYPASSWORD")
    static abstract class TestConfig {
        @Autowired
        protected Environment environment;
    }
}

src/test/java/eu/hinsch/spring/propertiesdecryptor/DecryptingPropertySourcesPlaceholderConfigurerTest.java
package eu.hinsch.spring.propertiesdecryptor;

import org.jasypt.exceptions.EncryptionOperationNotPossibleException;
import org.junit.After;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.test.SpringApplicationConfiguration;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.core.env.Environment;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.TestContextManager;
import org.springframework.test.context.support.DirtiesContextTestExecutionListener;

import java.util.ArrayList;
import java.util.List;

import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.isA;
import static org.junit.Assert.assertThat;
import static org.junit.internal.matchers.ThrowableCauseMatcher.hasCause;

/**
 * Created by lh on 04/04/15.
 */
@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
@SpringApplicationConfiguration(classes = DecryptingPropertySourcesPlaceholderConfigurerTest.TestConfig.class)
public class DecryptingPropertySourcesPlaceholderConfigurerTest {

    private static final String CODE = "Cn0fGsAEINM1+FsMf0evHob485wqmPwT";
    private static final String ENCRYPTED_VALUE = "{cypher}" + CODE;
    private static final String SECRET = "MY-SECRET";
    private List<String> systemProperties = new ArrayList<>();

    @Rule
    public ExpectedException exception = ExpectedException.none();

    @Autowired
    private Environment environment;
    @Autowired
    private ConfigurableApplicationContext applicationContext;

    private TestContextManager testContextManager;

    @SpringBootApplication
    static class TestConfig {
    }

    @Test
    public void shouldDecodeSystemProperty() throws Exception {
        // given
        setSystemProperty("secretProperty", ENCRYPTED_VALUE);

        // when
        startupContext();

        // then
        assertThat(environment.getProperty("secretProperty"), is(SECRET));
    }

    @Test
    public void shouldDecryptCorrectPropertyOnConflict() throws Exception {
        // given
        setSystemProperty("conflictingProperty", ENCRYPTED_VALUE);

        // when
        startupContext();

        //
        assertThat(environment.getProperty("conflictingProperty"), is(SECRET));
    }

    @Test
    public void shouldThrowExceptionIfPasswordIsIncorrect() throws Exception {
        // given
        setSystemProperty("propertyEncryption.password", "NOT-THE-CORRECT-PASSWORD");
        setSystemProperty("somethingToDecrypt", ENCRYPTED_VALUE);
        exception.expect(IllegalStateException.class);
        // meaningful exception is wrapped twice inside spring errors
        exception.expectCause(hasCause(isA(EncryptionOperationNotPossibleException.class)));

        // when
        startupContext();

        // then -> exception
    }

    @Test
    @Ignore("TODO not yet working")
    public void shouldUsePrefixOverride() throws Exception {
        // given
        setSystemProperty("propertyDecryption.prefix", "{cypher}");
        setSystemProperty("secretProperty", ENCRYPTED_VALUE);

        // when
        startupContext();

        // then
        assertThat(environment.getProperty("secretProperty"), is(SECRET));
    }

    private void setSystemProperty(String property, String value) {
        System.setProperty(property, value);
        systemProperties.add(property);
    }

    // manual app startup because we want to set properties before
    private void startupContext() throws Exception {
        testContextManager = new TestContextManager(this.getClass());
        testContextManager.prepareTestInstance(this);
        testContextManager.registerTestExecutionListeners(new DirtiesContextTestExecutionListener());
    }

    @After
    public void cleanUp() {
        systemProperties.stream().forEach(System::clearProperty);
    }
}

src/main/java/eu/hinsch/spring/propertiesdecrypter/DecryptingPropertySourcesPlaceholderConfigurer.java
package eu.hinsch.spring.propertiesdecrypter;

import org.jasypt.encryption.pbe.StandardPBEStringEncryptor;
import org.jasypt.salt.ZeroSaltGenerator;
import org.springframework.context.support.PropertySourcesPlaceholderConfigurer;
import org.springframework.core.env.*;

import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static java.util.stream.Collectors.toList;

/**
 * Created by lh on 02/04/15.
 */
public class DecryptingPropertySourcesPlaceholderConfigurer extends PropertySourcesPlaceholderConfigurer {
    static final String PASSWORD_PROPERTY = "propertyDecryption.password";
    public static final String DEFAULT_ALGORITHM = "PBEWithMD5AndDES";
    public static final String PREFIX_KEY = "propertyDecryption.prefix";

    private String prefix;
    private StandardPBEStringEncryptor encrypter = new StandardPBEStringEncryptor();

    @Override
    public void setEnvironment(final Environment environment) {
        initializeEncrypter(environment);
        prefix = environment.getProperty(PREFIX_KEY, "{encrypted}");

        final ConfigurableEnvironment configurableEnvironment = (ConfigurableEnvironment) environment;
        final MutablePropertySources propertySources = configurableEnvironment.getPropertySources();

        List<String> encryptedKeys = getKeysOfEncryptedPropertyValues(environment, propertySources);
        addDecryptedValues(environment, propertySources, encryptedKeys);

        super.setEnvironment(environment);
    }

    private void initializeEncrypter(Environment environment) {
        encrypter.setAlgorithm(environment.getProperty("propertyDecryption.algorithm", DEFAULT_ALGORITHM));
        encrypter.setSaltGenerator(new ZeroSaltGenerator());
        encrypter.setPassword(environment.getRequiredProperty(PASSWORD_PROPERTY));
    }

    private List<String> getKeysOfEncryptedPropertyValues(Environment environment, MutablePropertySources propertySources) {
        Stream<PropertySource<?>> stream = getPropertySourceStream(propertySources);
        return stream.filter(source -> source instanceof MapPropertySource)
                .map(source -> (MapPropertySource) source)
                .map(PropertySource::getSource)
                .flatMap(map -> map.entrySet().stream())
                .map(Map.Entry::getKey)
                .filter(key -> !PREFIX_KEY.equals(key))
                .filter(key -> isEncrypted(environment.getProperty(key)))
                .collect(toList());
    }

    private Stream<PropertySource<?>> getPropertySourceStream(final MutablePropertySources propertySources) {
        final Iterator<PropertySource<?>> iterator = propertySources.iterator();
        Iterable<PropertySource<?>> iterable = () -> iterator;
        return StreamSupport.stream(iterable.spliterator(), false);
    }

    private void addDecryptedValues(Environment environment, MutablePropertySources propertySources, List<String> encryptedKeys) {

        // TODO collect to properties?
        final Properties decryptedValues = new Properties();
        encryptedKeys.stream().forEach(
                key -> {
                    String value = environment.getProperty(key);
                    decryptedValues.put(key, decryptPropertyValue(value));
                }
        );
        propertySources.addFirst(new PropertiesPropertySource("decryptedValues", decryptedValues));
    }

    private String decryptPropertyValue(String encryptedPropertyValue) {
        String cypher = getCypher(encryptedPropertyValue);
        return encrypter.decrypt(cypher);
    }

    private boolean isEncrypted(Object propertyValue) {
        return propertyValue != null && propertyValue instanceof String && ((String)propertyValue).startsWith(prefix);
    }

    private String getCypher(String encryptedPropertyValue) {
        return encryptedPropertyValue.substring(prefix.length());
    }
}


src/main/java/eu/hinsch/spring/propertiesdecrypter/DecryptingPropertiesApplicationListener.java
package eu.hinsch.spring.propertiesdecrypter;

import org.jasypt.encryption.pbe.StandardPBEStringEncryptor;
import org.jasypt.exceptions.EncryptionOperationNotPossibleException;
import org.jasypt.salt.ZeroSaltGenerator;
import org.springframework.boot.context.event.ApplicationEnvironmentPreparedEvent;
import org.springframework.context.ApplicationListener;
import org.springframework.core.Ordered;
import org.springframework.core.env.*;

import java.util.Iterator;
import java.util.List;
import java.util.Properties;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static java.util.Arrays.asList;
import static java.util.stream.Collectors.toList;

/**
 * Created by lh on 02/04/15.
 */
public class DecryptingPropertiesApplicationListener
        implements ApplicationListener<ApplicationEnvironmentPreparedEvent>, Ordered {
    static final String PASSWORD_PROPERTY = "propertyDecryption.password";
    public static final String DEFAULT_ALGORITHM = "PBEWithMD5AndDES";
    public static final String PREFIX_KEY = "propertyDecryption.prefix";

    private String prefix;
    private StandardPBEStringEncryptor encrypter = new StandardPBEStringEncryptor();

    @Override
    public void onApplicationEvent(ApplicationEnvironmentPreparedEvent event) {
        ConfigurableEnvironment environment = event.getEnvironment();
        initializeEncrypter(environment);
        prefix = environment.getProperty(PREFIX_KEY, "{encrypted}");

        final MutablePropertySources propertySources = environment.getPropertySources();

        List<String> encryptedKeys = getKeysOfEncryptedPropertyValues(environment, propertySources);
        addDecryptedValues(environment, propertySources, encryptedKeys);
    }

    private void initializeEncrypter(Environment environment) {
        encrypter.setAlgorithm(environment.getProperty("propertyDecryption.algorithm", DEFAULT_ALGORITHM));
        encrypter.setSaltGenerator(new ZeroSaltGenerator());
        encrypter.setPassword(environment.getRequiredProperty(PASSWORD_PROPERTY));
    }

    private List<String> getKeysOfEncryptedPropertyValues(Environment environment, MutablePropertySources propertySources) {
        return getPropertySourceStream(propertySources)
                .filter(source -> source instanceof EnumerablePropertySource)
                .map(source -> (EnumerablePropertySource)source)
                .flatMap(source -> asList(source.getPropertyNames()).stream())
                .filter(this::isNotEncryptionConfigProperty)
                .filter(key -> isEncrypted(environment.getProperty(key)))
                .collect(toList());
    }

    private boolean isNotEncryptionConfigProperty(String key) {
        return !PREFIX_KEY.equals(key);
    }

    private Stream<PropertySource<?>> getPropertySourceStream(final MutablePropertySources propertySources) {
        final Iterator<PropertySource<?>> iterator = propertySources.iterator();
        Iterable<PropertySource<?>> iterable = () -> iterator;
        return StreamSupport.stream(iterable.spliterator(), false);
    }

    private void addDecryptedValues(Environment environment, MutablePropertySources propertySources, List<String> encryptedKeys) {
        // TODO collect to properties?
        final Properties decryptedValues = new Properties();
        encryptedKeys.stream()
                .forEach(key -> decryptAndAdd(environment, key, decryptedValues));
        propertySources.addFirst(new PropertiesPropertySource("decryptedValues", decryptedValues));
    }

    private void decryptAndAdd(Environment environment, String key, Properties decryptedValues) {
        final String property = environment.getProperty(key);
        final String propertyValue = decryptPropertyValue(property);
        decryptedValues.put(key, propertyValue);
    }

    private String decryptPropertyValue(String encryptedPropertyValue) {
        String cypher = getCypher(encryptedPropertyValue);
        try {
            return encrypter.decrypt(cypher);
        }
        catch (EncryptionOperationNotPossibleException e) {
            throw new RuntimeException("Unable to decrypt property value '" + encryptedPropertyValue + "'", e);
        }
    }

    private boolean isEncrypted(Object propertyValue) {
        return propertyValue != null && propertyValue instanceof String && ((String)propertyValue).startsWith(prefix);
    }

    private String getCypher(String encryptedPropertyValue) {
        return encryptedPropertyValue.substring(prefix.length());
    }

    @Override
    public int getOrder() {
        return Ordered.LOWEST_PRECEDENCE;
    }
}
